File: java/src/main/java/heaps/Twitter.java
package heaps;

import java.util.PriorityQueue;
import java.util.*;

public class Twitter {

    private int count = 0;
    private HashMap<Integer, HashSet<Integer>> followMap;
    private HashMap<Integer, List<int[]>> tweetMap; // userId -> List of [count, tweetIds]

    public Twitter() {
        followMap = new HashMap<>();
        tweetMap = new HashMap<>();
    }

    public void postTweet(int userId, int tweetId) {
        if (!tweetMap.containsKey(userId)) {
            tweetMap.put(userId, new ArrayList<>());
        }
        this.tweetMap.get(userId).add(new int[] {count, tweetId});
        count ++;
    }

    public List<Integer> getNewsFeed(int userId) {
        List<Integer> result = new ArrayList<>();
        java.util.PriorityQueue<int[]> maxHeap = new PriorityQueue<>((a, b) -> b[0] - a[0]);

        followMap.computeIfAbsent(userId, k -> new HashSet<>());

        followMap.get(userId).add(userId);

        for (int followeeId : followMap.get(userId)) {
            if (tweetMap.containsKey(followeeId)) {
                int i = tweetMap.get(followeeId).size() - 1;
                int[] tweet = tweetMap.get(followeeId).get(i);
                maxHeap.offer(new int[]{tweet[0], tweet[1], followeeId, --i});
            }
        }

        while (!maxHeap.isEmpty() && result.size() < 10) {
            int[] data = maxHeap.poll();
            result.add(data[1]);

            if (data[3] >= 0) {
                int[] tweet = tweetMap.get(data[2]).get(data[3]);
                maxHeap.offer(new int[]{tweet[0], tweet[1], data[2], --data[3]});
            }
        }

        return result;
    }

    public void follow(int followerId, int followeeId) {
        if (!followMap.containsKey(followerId)) {
            followMap.put(followerId, new HashSet<>());
        }

        followMap.get(followerId).add(followeeId);
    }

    public void unfollow(int followerId, int followeeId) {
        if (followMap.containsKey(followerId)&& followMap.get(followerId).contains(followeeId))
            followMap.get(followerId).remove(followeeId);
    }

    public static void main(String[] args) {
        Twitter twitter = new Twitter();
        twitter.postTweet(1, 5);
        twitter.getNewsFeed(1);

        twitter.follow(1, 2);
        twitter.postTweet(2, 6);
        twitter.getNewsFeed(1);
        twitter.unfollow(1, 2);
        twitter.getNewsFeed(1);
    }
}

File: python/heaps/Twitter.py
from typing import List
from collections import defaultdict
import heapq
# {
#     1: {}
# }
# {
#     1: [[tweetId, count], [tweetId, count]]
# }
class Twitter:

    def __init__(self):
        self.count = 0
        self.followMap = defaultdict(set)
        self.tweetMap = defaultdict(list)

    def postTweet(self, userId: int, tweetId: int) -> None:
        self.tweetMap[userId].append([self.count, tweetId])
        self.count -= 1

    def getNewsFeed(self, userId: int) -> List[int]:
        res = [] #List of 10 most recent tweet Ids

        heap = []

        for followeeId in self.followMap[userId]:
            if self.tweetMap[followeeId]:
                index = len(self.tweetMap[followeeId]) - 1
                count, tweetId = self.tweetMap[followeeId][index]
                heap.append([count, tweetId, followeeId, index - 1])
        heapq.heapify(heap)
        
        while heap and len(res) <= 10:
            count, tweetId, followeeId, index = heapq.heappop(heap)
            res.append(tweetId)

            if index >= 0:
                count, tweetId = self.tweetMap[followeeId][index]
                heapq.heappush(heap, [count, tweetId, followeeId, index - 1])
    
        return res
        
    def follow(self, followerId: int, followeeId: int) -> None:
        self.followMap[followerId].add(followeeId)

    def unfollow(self, followerId: int, followeeId: int) -> None:
        if (self.followMap[followerId] and 
            followeeId in self.followMap[followerId]):
            self.followMap[followerId].remove(followeeId)

obj = Twitter() #inst
obj.follow(1, 5)
obj.getNewsFeed(1)
obj.follow(1, 2)
obj.postTweet(2, 6)
obj.postTweet(2, 7)
obj.postTweet(5, 8)
print(obj.getNewsFeed(1))
